File: source/sdks/sd-jwt-sdk/src/main/java/org/omnione/did/sdjwt/datamodel/SDJWT.java
/*
 * Copyright 2025 OmniOne.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.omnione.did.sdjwt.datamodel;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import org.omnione.did.sdjwt.exception.SDJWTException;

//TODO: Structured SD-JWT must be supported.
public class SDJWT {

  private final String credentialJwt;
  private final List<Disclosure> disclosures;
  private final String keyBindingJwt;

  public SDJWT(String credentialJwt, Collection<Disclosure> disclosures) {
    this(credentialJwt, disclosures, null);
  }

  public SDJWT(String credentialJwt, Collection<Disclosure> disclosures, String keyBindingJwt) {
    if (credentialJwt == null || credentialJwt.trim().isEmpty()) {
      throw new IllegalArgumentException("Credential JWT cannot be null or empty");
    }

    this.credentialJwt = credentialJwt.trim();
    this.disclosures = disclosures != null ?
        new ArrayList<>(disclosures) : new ArrayList<>();
    this.keyBindingJwt = keyBindingJwt != null && !keyBindingJwt.trim().isEmpty() ?
        keyBindingJwt.trim() : null;
  }

  public String getCredentialJwt() {
    return credentialJwt;
  }

  public List<Disclosure> getDisclosures() {
    return Collections.unmodifiableList(disclosures);
  }

  public String getKeyBindingJwt() {
    return keyBindingJwt;
  }

  public boolean hasKeyBindingJwt() {
    return keyBindingJwt != null;
  }

  public int getDisclosureCount() {
    return disclosures.size();
  }

  public boolean hasDisclosures() {
    return !disclosures.isEmpty();
  }

  public static SDJWT parse(String sdJwtString) {
    if (sdJwtString == null || sdJwtString.trim().isEmpty()) {
      throw new SDJWTException("SD-JWT string cannot be null or empty");
    }

    String[] parts = sdJwtString.split("~", -1);

    if (parts.length < 1) {
      throw new SDJWTException("SD-JWT must contain at least a credential JWT");
    }

    String credentialJwt = parts[0];
    if (credentialJwt.isEmpty()) {
      throw new SDJWTException("Credential JWT cannot be empty");
    }

    if (!isValidJwtFormat(credentialJwt)) {
      throw new SDJWTException("Invalid credential JWT format");
    }

    List<Disclosure> disclosures = new ArrayList<>();
    String keyBindingJwt = null;

    for (int i = 1; i < parts.length; i++) {
      String part = parts[i];

      if (part.isEmpty()) {

        continue;
      }

      if (isValidJwtFormat(part)) {

        if (i == parts.length - 1 || areAllEmpty(parts, i + 1)) {
          keyBindingJwt = part;
        } else {
          throw new SDJWTException("Key binding JWT can only be the last component");
        }
      } else {

        try {
          Disclosure disclosure = Disclosure.parse(part);
          disclosures.add(disclosure);
        } catch (SDJWTException e) {
          throw e;
        }
      }
    }

    return new SDJWT(credentialJwt, disclosures, keyBindingJwt);
  }

  private static boolean areAllEmpty(String[] parts, int fromIndex) {
    for (int i = fromIndex; i < parts.length; i++) {
      if (!parts[i].isEmpty()) {
        return false;
      }
    }
    return true;
  }

  private static boolean isValidJwtFormat(String jwt) {
    if (jwt == null || jwt.trim().isEmpty()) {
      return false;
    }

    String[] parts = jwt.split("\\.");
    return parts.length == 3 &&
        !parts[0].isEmpty() &&
        !parts[1].isEmpty();
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();

    sb.append(credentialJwt);

    for (Disclosure disclosure : disclosures) {
      sb.append("~").append(disclosure.getDisclosure());
    }

    if (keyBindingJwt != null) {
      sb.append("~").append(keyBindingJwt);
    } else {

      sb.append("~");
    }

    return sb.toString();
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    SDJWT sdjwt = (SDJWT) o;
    return Objects.equals(credentialJwt, sdjwt.credentialJwt) &&
        Objects.equals(disclosures, sdjwt.disclosures) &&
        Objects.equals(keyBindingJwt, sdjwt.keyBindingJwt);
  }

  @Override
  public int hashCode() {
    return Objects.hash(credentialJwt, disclosures, keyBindingJwt);
  }
}

File: source/apps/ios-app/oid4vc/oid4vc/Models/SDJWT.swift
//
/*
 * Copyright 2026 OmniOne.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
    
import Foundation
import CryptoKit

// MARK: - SD-JWT minimal models/utilities (used by CredentialMatcher)

public struct SDJWT : Jsonable {
    public let credentialJwt: String
    public let disclosures: [Disclosure]
    public var keyBindingJwt: String?
    
    public init(credentialJwt: String, disclosures: [Disclosure] = [], keyBindingJwt: String? = nil) {
        self.credentialJwt = credentialJwt
        self.disclosures   = disclosures
        self.keyBindingJwt = keyBindingJwt
    }
    
    public static func parse(raw: String) -> SDJWT
    {
        // SD-JWT is formatted as: credentialJwt~disclosure1~disclosure2~...~[keyBindingJwt]
        // Split by '~' preserving order
        let parts = raw.components(separatedBy: "~")
        guard !parts.isEmpty else {
            return SDJWT(credentialJwt: raw, disclosures: [], keyBindingJwt: nil)
        }

        // First segment is the credential JWT
        let credentialJwt = parts[0]

        // Determine if there is a keyBindingJwt by counting dots in the entire raw string
        // Typical JWT has 2 dots. If there are 4 dots overall, then a key binding JWT exists at the end.
        let hasKeyBinding = raw.getCount(of: ".") >= 4

        // Disclosures are any middle segments that are valid disclosure encodings
        let startIndexForDisclosures = 1
        let endIndexForDisclosuresExclusive = hasKeyBinding ? (parts.count - 1) : parts.count
        let disclosureSegments: [String]
        if startIndexForDisclosures < endIndexForDisclosuresExclusive {
            disclosureSegments = Array(parts[startIndexForDisclosures..<endIndexForDisclosuresExclusive])
        } else {
            disclosureSegments = []
        }

        let disclosures = disclosureSegments.compactMap { Disclosure.parse(raw: $0) }

        // Optional key binding JWT is the last segment if present
        let keyBindingJwt: String? = hasKeyBinding ? parts.last : nil

        return SDJWT(credentialJwt: credentialJwt, disclosures: disclosures, keyBindingJwt: keyBindingJwt)
    }
    
    public func toString() -> String {
        
        var jwt = credentialJwt
        
        for disclosure in disclosures {
            jwt.append("~\(disclosure.getDisclosure())")
        }
        
        jwt.append("~")
        
        if let keyBinding = keyBindingJwt
        {
            jwt.append(keyBinding)
        }
        
        return jwt
    }
}

extension String
{
    func getCount(of word: String) -> Int
    {
        guard !word.isEmpty else { return 0 }

        var count = 0
        var searchRange: Range<String.Index>? = startIndex..<endIndex

        while let range = self.range(of: word, range: searchRange) {
            count += 1
            searchRange = range.upperBound..<endIndex
        }

        return count
    }
}

public struct Disclosure: Codable, Equatable {
    public let salt: String
    public let claimName: String?
    public let claimValue: JSON

    public init(salt: String, claimName: String?, claimValue: JSON) {
        self.salt = salt
        self.claimName = claimName
        self.claimValue = claimValue
    }

    // JSON array payload (sorted keys for stable digest)
    func toData() -> Data {
        var arr: [JSON] = [.string(salt)]
        if let claimName = claimName {
            arr.append(.string(claimName))
        }
        arr.append(claimValue)

        let encoder = RNJSONEncoder()
        return (try? encoder.encode(arr)) ?? Data()
    }

    public func getDisclosure() -> String {
        self.toData().base64URLEncodedString()
    }

    public func digest() -> String {
        let hash = SHA256.hash(data: self.toData())
        return Base64URL.encode(Data(hash))
    }

    /// Parse SD-JWT disclosure (base64url JSON array)
    /// Typical forms:
    /// - ["salt","claimName", claimValue]
    /// - ["salt", claimValue]  (array item disclosure)
    public static func parse(raw: String) -> Disclosure? {
        guard let decoded = Base64URL.decode(raw) else { return nil }

        let top: JSON
        do {
            top = try JSONParser().parse(data: decoded)
        } catch {
            return nil
        }

        guard case let .array(items) = top, items.count >= 2 else { return nil }

        guard case let .string(salt) = items[0] else { return nil }

        if items.count == 3 {
            guard case let .string(claimName) = items[1] else { return nil }
            let claimValue = items[2]
            return Disclosure(salt: salt, claimName: claimName, claimValue: claimValue)
        }

        if items.count == 2 {
            let claimValue = items[1]
            return Disclosure(salt: salt, claimName: nil, claimValue: claimValue)
        }

        return nil
    }
}
